keps/sig-node/xxxx-container-oom-kill-mode/README.md

```diff
@@ -0,0 +1,729 @@
+
+## Design Details
+
+### API Changes
+
+#### Container API
+
+The changes will be made to the Container struct in `staging/src/k8s.io/api/core/v1/types.go`:
+
+```go
+// Container represents a single container that is expected to be run on the host.
+type Container struct {
+    // ... existing fields ...
+
+    // OOMKillMode specifies how the OOM killer behaves for this container.
+    // +featureGate=ContainerOOMKillMode
+    // +optional
+    OOMKillMode *OOMKillMode `json:"oomKillMode,omitempty" protobuf:"bytes,26,opt,name=oomKillMode"`
+}
+```
+
+The same field applies to `EphemeralContainer` for consistency.
+
+### Kubelet Implementation
+
+The kubelet will be modified in `pkg/kubelet/kuberuntime/kuberuntime_container_linux.go` to respect the container-level setting:
+
+```go
+func (m *kubeGenericRuntimeManager) generateLinuxContainerResources(
+    ctx context.Context,
+    pod *v1.Pod,
+    container *v1.Container,
+    enforceMemoryQoS bool,
+) *runtimeapi.LinuxContainerResources {
+    // ... existing code ...
+
+    // v1.36+: Warn about deprecated flag
+    if utilfeature.DefaultFeatureGate.Enabled(features.ContainerOOMKillMode) &&
+       m.singleProcessOOMKill != nil {
+        klog.Warning("Flag --single-process-oom-kill is deprecated and will be removed in v1.38. " +
+                    "Please migrate to pod.spec.containers.oomKillMode field.")
+    }
+
+    // Determine OOM kill behavior
+    var useGroupKill bool
+
+    if container.OOMKillMode != nil {
+        // Container-level setting takes precedence
+        switch *container.OOMKillMode {
+        case v1.OOMKillModeSingle:
+            useGroupKill = false
+        case v1.OOMKillModeGroup:
+            if !isCgroup2UnifiedMode() {
+                klog.Warningf("Container %s/%s requests Group mode but system uses cgroup v1, falling back to single process",
+                    pod.Name, container.Name)
+                useGroupKill = false
+            } else {
+                useGroupKill = true
+            }
+        }
+    } else {
+        // Fall back to kubelet configuration (v1.35-v1.37)
+        // After v1.38, default to cgroup v2 behavior (group kill)
+        if utilfeature.DefaultFeatureGate.Enabled(features.ContainerOOMKillMode) &&
+           !utilfeature.DefaultFeatureGate.Enabled(features.LegacyKubeletFlags) {
+            // v1.38+: Flag removed, use cgroup default
+            useGroupKill = isCgroup2UnifiedMode()
+        } else {
+            // v1.35-v1.37: Still support the flag
+            useGroupKill = isCgroup2UnifiedMode() && !ptr.Deref(m.singleProcessOOMKill, false)
+        }
+    }
+
+    // Apply the configuration
+    if isCgroup2UnifiedMode() && useGroupKill {
+        resources.Unified = map[string]string{
+            "memory.oom.group": "1",
+        }
+    }
+
+    return &resources
+}
+```
+
+### Validation Rules
+
+API validation will be added in `pkg/apis/core/validation/validation.go`:
```

Comment:
If Pod spec's `OS` field is set to windows, it might be a better to prevent users from specifying `OOMKillMode`.
https://github.com/kubernetes/kubernetes/blob/v1.33.4/pkg/apis/core/types.go#L3680